src/java/org/apache/cassandra/db/memtable/AbstractMemtable.java

```diff
@@ -0,0 +1,188 @@
+import java.util.Map;
+import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.cassandra.db.RegularAndStaticColumns;
+import org.apache.cassandra.db.commitlog.CommitLogPosition;
+import org.apache.cassandra.db.partitions.Partition;
+import org.apache.cassandra.db.rows.EncodingStats;
+import org.apache.cassandra.schema.ColumnMetadata;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.schema.TableMetadataRef;
+import org.apache.cassandra.utils.concurrent.OpOrder;
+
+public abstract class AbstractMemtable implements Memtable
+{
+    protected final AtomicLong liveDataSize = new AtomicLong(0);
+    protected final AtomicLong currentOperations = new AtomicLong(0);
+    protected final ColumnsCollector columnsCollector;
+    protected final StatsCollector statsCollector = new StatsCollector();
+    // The smallest timestamp for all partitions stored in this memtable
+    protected AtomicLong minTimestamp = new AtomicLong(Long.MAX_VALUE);
+    protected TableMetadataRef metadata;
+
+    public AbstractMemtable(TableMetadataRef metadataRef)
+    {
+        this.metadata = metadataRef;
+        this.columnsCollector = new ColumnsCollector(metadata.get().regularAndStaticColumns());
+    }
+
+    public TableMetadata metadata()
+    {
+        return metadata.get();
+    }
+
+    public long getLiveDataSize()
+    {
+        return liveDataSize.get();
+    }
+
+    public long getOperations()
+    {
+        return currentOperations.get();
+    }
+
+    public long getMinTimestamp()
+    {
+        return minTimestamp.get();
+    }
+
+    protected void updateMin(AtomicLong minTracker, long newValue)
+    {
+        while (true)
+        {
+            long memtableMinTimestamp = minTracker.get();
+            if (memtableMinTimestamp <= newValue)
+                break;
+            if (minTracker.compareAndSet(memtableMinTimestamp, newValue))
+                break;
+        }
+    }
+
+    RegularAndStaticColumns columns()
+    {
+        return columnsCollector.get();
+    }
+
+    EncodingStats encodingStats()
+    {
+        return statsCollector.get();
+    }
+
+    protected static class ColumnsCollector
+    {
+        private final HashMap<ColumnMetadata, AtomicBoolean> predefined = new HashMap<>();
+        private final ConcurrentSkipListSet<ColumnMetadata> extra = new ConcurrentSkipListSet<>();
+
+        ColumnsCollector(RegularAndStaticColumns columns)
+        {
+            for (ColumnMetadata def : columns.statics)
+                predefined.put(def, new AtomicBoolean());
+            for (ColumnMetadata def : columns.regulars)
+                predefined.put(def, new AtomicBoolean());
+        }
+
+        public void update(RegularAndStaticColumns columns)
+        {
+            for (ColumnMetadata s : columns.statics)
```

Comment:
Correctness wrt concurrency is not trivial in this class I think:

1. `update()` and `get()` may be interleaved, correct? (streaming case); if this is incorrect let's have explicit assertion that it doesn't happen   
2. if they can be interleaved then the `get()` caller may be a somewhat unexpected view of the columns (E.g. may see a newer update, but miss an older one). I _assume_ it is OK

Could we state 2 explicitly in some comment?

Reply:
Added statement to that effect.

This, or more precisely the fact that the gathered collections can be out-of-date with respect to data that is being flushed, can be a problem for persistent metadata. This is not trivial to fix; added comments stating the problem.